### PR TITLE
Feat: recurrent mappo vmapped agents

### PR DESCRIPTION
**Pull request by OmaymaMahjoub**
## What?
First fully end to end JAX recurrent MAPPO Agent that is able to train on Rware scenarios. 

## Note
This is joint work done with @RuanJohn.

### CODE REVIEW

File: jax_distribution/evaluator.py [hunk 1]
```diff
@@ -138,35 +125,212 @@ def evaluator_fn(trained_params: FrozenDict, rng: chex.PRNGKey) -> ExperimentOut
...
-    )
+    # Vmap it over number of agents and create evaluator_fn.
+    if rnn_net:
+        if centralised_critic:
+            vmapped_eval_network_apply_fn = jax.vmap(
+                network.apply,
+                in_axes=(
+                    None,
+                    (1, None),
+                    (ObservationGlobalState(2, 2, None, 2), None),
```
Comment (sash-a) [inline jax_distribution/evaluator.py:295]: Howcome you don't need to do this if we're ff centralised?

File: jax_distribution/logger.py [hunk 2]
```diff
@@ -19,26 +19,24 @@
...
 from sacred.run import Run
 
-from jax_distribution.types import ExperimentOutput
 from jax_distribution.utils.logger_tools import Logger
 
 
-# todo: flake is complaining, possibly split into 3 functions
-#  log_absolute, log_trainer, log_evaluator
-def get_logger_fn(logger: Logger, config: Dict) -> Callable:  # noqa: CCR001
+def get_logger_fn(logger: SacredLogger, config: Dict) -> Callable:  # noqa: CCR001
```
Comment (sash-a) [inline jax_distribution/logger.py:31]: I think revert this file as the old stuff is correct, assume it is still being merged

File: jax_distribution/systems/rec_mappo_rware.py [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+
+            rng, policy_rng = jax.random.split(rng)
+
+            # Add a batch dimension to the observation.
+            batched_observation = jax.tree_util.tree_map(
+                lambda x: x[np.newaxis, :], last_timestep.observation
+            )
+            ac_in = (
+                batched_observation,
+                last_done[:, 0][np.newaxis, :],
```
Comment (sash-a) [inline jax_distribution/systems/rec_mappo_rware.py:198]: Only seeing this now, but I think in other places we use `None` to add a new axis, I actually quite like `np.newaxis` as it's clearer what is happening, but we should just stay consistent, so either change these to `None` or change the others to `np.newaxis`

File: jax_distribution/systems/rec_mappo_rware.py [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+                    # CALCULATE ACTOR LOSS
+                    ratio = jnp.exp(log_prob - traj_batch.log_prob)
+                    gae = (gae - gae.mean()) / (gae.std() + 1e-8)
+                    loss_actor1 = ratio * gae
+                    loss_actor2 = (
+                        jnp.clip(
+                            ratio,
+                            1.0 - config["clip_eps"],
+                            1.0 + config["clip_eps"],
+                        )
```
Comment (sash-a) [inline jax_distribution/systems/rec_mappo_rware.py:357]: Also only saw this now, would need to change all files, but might be better to pull this out into `clipped_ratio = ...` for readability

File: jax_distribution/systems/rec_mappo_rware.py [hunk 5]
```diff
@@ -0,0 +1,737 @@
...
+                init_critic_hstate,
+                traj_batch,
+                advantages,
+                targets,
+            )
+            shuffled_batch = jax.tree_util.tree_map(
+                lambda x: jnp.take(x, permutation, axis=1), batch
+            )
+            minibatches = jax.tree_util.tree_map(
+                lambda x: jnp.swapaxes(
```
Comment (sash-a) [inline jax_distribution/systems/rec_mappo_rware.py:413]: As commented on previous PR pull out the `reshape`

File: jax_distribution/systems/rec_mappo_rware.py (hunk 3 above)
Comment (sash-a) [reply jax_distribution/systems/rec_mappo_rware.py:198]: Having gone through the PR I don't see adding a new axis with `None` anywhere, so happy for you to keep things as they are. Just have a look in the other files as I feel like I saw it somewhere :shrug:

**Review by sash-a — COMMENTED** [review]
Nice looks great, some more very minor stuff! Just commenting so I don't block the merge later. Good luck!! :fire:

File: jax_distribution/logger.py (hunk 2 above)
Comment (sash-a) [reply jax_distribution/logger.py:31]: Seems like there was a mix up in the merge, definitely revert this file

File: jax_distribution/evaluator.py (hunk 1 above)
Comment (OmaymaMahjoub) [reply jax_distribution/evaluator.py:295]: We already did the same for the mappo check this[ line](https://github.com/instadeepai/Mava/blob/feat/pure-jax-mava/jax_distribution/systems/ff_mappo_rware.py#L355)

File: jax_distribution/systems/rec_mappo_rware.py (hunk 3 above)
Comment (OmaymaMahjoub) [reply jax_distribution/systems/rec_mappo_rware.py:184]: We changed them to jnp.newaxis

File: jax_distribution/systems/rec_mappo_rware.py (hunk 4 above)
Comment (OmaymaMahjoub) [reply jax_distribution/systems/rec_mappo_rware.py:333]: We feel like this is already descriptive enough

Comment (OmaymaMahjoub) [reply jax_distribution/systems/rec_mappo_rware.py:333]: are you okay to reconsider

File: jax_distribution/systems/rec_mappo_rware.py [hunk 6]
```diff
@@ -409,17 +382,14 @@ def _loss_fn(
...
-                    jnp.reshape(
-                        x,
-                        [x.shape[0], config["num_minibatches"], -1] + list(x.shape[2:]),
-                    ),
-                    1,
-                    0,
+            reshaped_batch = jax.tree_util.tree_map(
+                lambda x: jnp.reshape(
+                    x,
+                    [x.shape[0], config["num_minibatches"], -1] + list(x.shape[2:]),
```
Comment (sash-a) [inline jax_distribution/systems/rec_mappo_rware.py:388]: Can we change this slightly, here and in the other PR::

```suggestion
                    x, (x.shape[0], config["num_minibatches"], -1, *x.shape[2:])
```

**Review by sash-a — APPROVED** [review]
Nice! One very nitpicky change, could you also do it in the other files please :smile:

**Review by RuanJohn — APPROVED** [review]
THis looks good to me! Thank you @OmaymaMahjoub! :fire:
